core/src/main/java/org/elasticsearch/search/fetch/subphase/VersionFetchSubPhase.java

```diff
@@ -18,32 +18,45 @@
  */
 package org.elasticsearch.search.fetch.subphase;
 
+import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.NumericDocValues;
-import org.elasticsearch.ElasticsearchException;
+import org.apache.lucene.index.ReaderUtil;
 import org.elasticsearch.common.lucene.uid.Versions;
 import org.elasticsearch.index.mapper.VersionFieldMapper;
+import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.fetch.FetchSubPhase;
 import org.elasticsearch.search.internal.SearchContext;
 
 import java.io.IOException;
+import java.util.Arrays;
+import java.util.Comparator;
 
 public final class VersionFetchSubPhase implements FetchSubPhase {
-
     @Override
-    public void hitExecute(SearchContext context, HitContext hitContext) {
+    public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOException {
         if (context.version() == false ||
             (context.storedFieldsContext() != null && context.storedFieldsContext().fetchFields() == false)) {
             return;
         }
-        long version = Versions.NOT_FOUND;
-        try {
-            NumericDocValues versions = hitContext.reader().getNumericDocValues(VersionFieldMapper.NAME);
-            if (versions != null && versions.advanceExact(hitContext.docId())) {
+
+        hits = hits.clone(); // don't modify the incoming hits
```

Comment:
I wonder if we can have some kind of utility for this kind of logic?

Something like:

`decorateSearchHit(SearchHits hits, BiFunction<String, LeafReaderContext, DocValuesIterator> dvLoader, BiConsumer<SearchHit, DocValuesIterator> decorateFunction) { `

So that we don't have to repeat the logic for cloning search hits and keeping track of reader id / dv instance?

Reply:
I think we should discuss such ideas in follow-up PRs. It's not clear to me that replacing duplicate code with more abstractions would be a win.

Reply:
Sure, I was just wondering as this patterns appears now at least 3 times.